nixos/modules/services/monitoring/grafana.nix

```diff
@@ -50,6 +50,166 @@ let
     SMTP_FROM_ADDRESS = cfg.smtp.fromAddress;
   } // cfg.extraOptions;
 
+  datasourceConfiguration = {
+    apiVersion = 1;
+    datasources = cfg.provision.datasources;
+  };
+
+  datasourceFile = pkgs.writeText "datasource.yaml" (builtins.toJSON datasourceConfiguration);
+
+  dashboardConfiguration = {
+    apiVersion = 1;
+    providers = cfg.provision.dashboards;
+  };
+
+  dashboardFile = pkgs.writeText "dashboard.yaml" (builtins.toJSON dashboardConfiguration);
+
+  provisionConfDir = pkgs.stdenv.mkDerivation {
+    name = "grafana-provisioning";
+    buildCommand = ''
+      mkdir $out
+      mkdir $out/datasources
+      mkdir $out/dashboards
+      ln -sf ${datasourceFile} $out/datasources/datasource.yaml
+      ln -sf ${dashboardFile} $out/dashboards/dashboard.yaml
+    '';
+  };
+
+  # Get a submodule without any embedded metadata:
+  _filter = x: filterAttrs (k: v: k != "_module") x;
+
+  # http://docs.grafana.org/administration/provisioning/#datasources
+  grafanaTypes.datasourceConfig = types.submodule {
+    options = {
+      name = mkOption {
+        type = types.str;
+        default = "Grafana";
```

Comment:
Naming a grafana datasource "Grafana" is not a good idea IMHO. As this field is required in the UI as well, I wouldn't set a default here.